core/src/display_object/movie_clip.rs

```diff
@@ -1080,15 +1090,45 @@ impl<'gc> MovieClip<'gc> {
         }
     }
 
+    /// Advance playhead to the next frame in the clip.
+    ///
+    /// This supports three different actions as stated in `NextFrame`. The
+    /// selected action is stored in `natural_playhead_action` and can be
+    /// referenced at different parts during frame execution.
+    fn advance_playhead(self, context: &mut UpdateContext<'_, 'gc, '_>) -> NextFrame {
+        let playhead_action = self.determine_next_frame();
+        self.0.write(context.gc_context).natural_playhead_action = playhead_action;
+
+        match playhead_action {
+            NextFrame::Next => {
+                let mut write = self.0.write(context.gc_context);
+
+                write.queued_script_frame = Some(write.current_frame + 1);
+                write.current_frame += 1;
+            }
+            NextFrame::First => self.run_goto(context, 1, true),
+            NextFrame::Same => self.stop(context),
+        };
+        playhead_action
+    }
+
     fn run_frame_internal(
         self,
         context: &mut UpdateContext<'_, 'gc, '_>,
         run_display_actions: bool,
     ) {
-        match self.determine_next_frame() {
-            NextFrame::Next => self.0.write(context.gc_context).current_frame += 1,
-            NextFrame::First => return self.run_goto(context, 1, true),
-            NextFrame::Same => self.stop(context),
+        let vm_type = context.avm_type();
+        let action = if vm_type == AvmType::Avm1 {
+            self.advance_playhead(context)
+        } else {
+            self.0.read().natural_playhead_action
+        };
+
+        //NOTE: For some reason, we actually have to use the playhead action
+        //that `advance_playhead` returns. If you pull `natural_playhead_action`
+        //out of the clip, you get some weird miscompile and AVM1 breaks
```

Comment:
you mean, Rust miscompile? Does this still happen on latest nightly/stable?

Reply:
Broken on 1.58 (released 2 days ago); just reading `self.0.read().natural_playhead_action` and not using the return on `advance_playhead` breaks the `default_names` test.